### DISCUSSION TITLE
How can I number all callouts by default

### DISCUSSION THREAD

**Discussion by normansimonr** [quarto-dev/quarto-cli#10240] [Q&A] [answered]
### Description

I know that if I add an identifier to a callout, it will be numbered and referenceable (as per the [docs](https://quarto.org/docs/authoring/cross-references.html#callouts)).

I have a document with an extremely large number of callouts, but only some of them are actually referenced (linked to) in other parts of the document. I want all of the callouts, however, including those that are referenced and those that aren't, to be numbered by default. How can I achieve this without having to manually add identifiers to all of the callouts?

Thanks.

**Comment by mcanouil** [comment]
You're gonna need to write a LUA filter that will add the identifiers.

**Comment by cscheid** [comment]
This filter will be slightly tricky to write because a Callout is a custom AST. So here's how you do it:

## doc.qmd

````
---
format: html
filters:
  - addcalloutid.lua
---

::: {#nte-1 .callout-note}

This is note 1.

:::

::: {.callout-note}

This is note 2.

:::

::: {.callout-note}

This is note 3.

:::

::: {.callout-note}

This is note 4.

:::
````

## `addcalloutid.lua`

```lua
local id = 1
function Callout(callout)
  if callout.attr.identifier ~= "" then
    -- callout already has an id
    return nil
  end
  callout.attr.identifier = "nte-unique" .. tostring(id)
  id = id + 1
  return quarto.Callout(callout)
end
```

This will add ids `nte-unique1`, `nte-unique2`, and `nte-unique3` to the last three callouts of the document, producing this output:

<img width="935" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/3de16749-fd20-433a-9f6a-fcf72453ac71">

**Reply by mcanouil** [reply]
The call to the Call-out constructor is mandatory? I wrote a similar filter but was not able to make it work so looking at your code it seems the constructor was the missing piece.


---

For future readers:

To have the same thing for other call-out types, `nte` needs to be replaced (or extend the filter to account for other prefixes) by one of the other prefix listed in <https://quarto.org/docs/authoring/callouts.html#cross-references>.

**Reply by cscheid** [reply]
Yes, it is, because callouts are custom AST nodes. (The same thing would be necessary if filters were to create crossreferenceable floats)

**Reply by mcanouil** [reply]
Thanks, so definitely the missing piece.
I assume that all custom nodes in Quarto would need to use the constructor, am I correct?

**Reply by cscheid** [reply]
For most definitions of "all custom nodes", yes :) We have some custom AST nodes that we only use internally.

**Reply by coatless** [reply]
Let's say I have a custom `todo` callout. To setup a prefix, I would only need to set `callout.attr.identifier`? 

So, we would only need to switch:

```lua
callout.attr.identifier = "nte-unique-" .. tostring(id)
```

To:

```lua
callout.attr.identifier = "tdo-unique-" .. tostring(id)
```

Then, the first todo could be referenced with `@tdo-unique-1`?

**Reply by cscheid** [reply]
Can we follow up offline? I think it would be easier to clarify all of this over email and then post the results here, rather than attempt to diagnose/fix everything online and potentially confusing everyone else. Shoot me an email and I'd be happy to chat. Thanks!

**Reply by coatless** [reply]
@cscheid reach out via the gmail address listed on your website?

**Reply by cscheid** [reply]
I thought I had my posit email on my github account, but I don't. The gmail one works, but carlos.scheidegger@posit.co is my work address.

**Reply by coatless** [reply]
Okidokies.

**Reply by coatless** [reply]
@cscheid was anything ever codified regarding setting up custom ast references? I had an inquiry come over:

https://github.com/coatless-quarto/custom-callout/issues/10

My github search-fu indicates no change; but, I know a new parser is in the works so...

**Comment by normansimonr** [comment]
Thanks so much @cscheid and @mcanouil! I'm new to extensions, so I'm following the docs to get started, but apparently I was able to make it work. The problem I'm getting is that all the notes have the same number, namely 'Note 92'. I copied the code verbatim. I'm working on a LaTeX doc.

I'm guessing the id needs to be updated at some point in a loop or something like that?

**Reply by cscheid** [reply]
@normansimonr Huh. No, that's a Quarto bug. Thanks for reporting, I can repro on that very example I shared with you. I'll follow up.

**Reply by cscheid** [reply]
For completeness, the example I shared produces this in `format: pdf`:

<img width="711" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/c47dd2aa-a968-41e6-b964-ac89ba7d5179">

**Reply by cscheid** [reply, marked as answer]
Oh, I'm sorry, it's not a Quarto bug (whew!), it's a simple copy-paste mistake I made. The correct filter is this, it just needs to increment that counter:

````lua
local id = 1
function Callout(callout)
  if callout.attr.identifier ~= "" then
    -- callout already has an id
    return nil
  end
  callout.attr.identifier = "nte-unique" .. tostring(id)
  id = id + 1
  return quarto.Callout(callout)
end
````

**Reply by cscheid** [reply]
<img width="753" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/f3202b22-9e79-48c6-8e35-3093abc2f67b">

**Reply by pglpm** [reply]
> ```lua
> local id = 1
> function Callout(callout)
>   if callout.attr.identifier ~= "" then
>     -- callout already has an id
>     return nil
>   end
>   callout.attr.identifier = "nte-unique" .. tostring(id)
>   id = id + 1
>   return quarto.Callout(callout)
> end
> ```

This is very useful for me too! One question: is it possible to change it so that it only applies to one type of callout, say the "caution" one? I've tried inserting an `if` with `callout.type`, but probably it doesn't exist.

**Reply by mcanouil** [reply]
Yes, test the `type` which should work, if not please share a small but fully reproducible example.

https://quarto.org/docs/prerelease/1.3/custom-ast-nodes/callout.html

---

#### Edit

It does exist and works:

````lua
function Callout(callout)
  quarto.log.output(callout.type)
end
````

````qmd
---
format: html
filters:
  - path: callout.lua
    at: post-quarto
---

::: {.callout-important}
Whatever
:::
````

**Reply by pglpm** [reply]
My bad, I was using `=` instead of `==` for matching. Thank you so much!

For other users who are interested in numbering only a certain type of callouts, say "caution":

```{lua}
local id = 1
function Callout(callout)
  if callout.type == "caution" then
    if callout.attr.identifier ~= "" then
      -- callout already has an id
      return nil
    end
    callout.attr.identifier = "cau-unique" .. tostring(id)
    id = id + 1
  end
  return quarto.Callout(callout)
end
```

**Comment by normansimonr** [comment]
Got it, thanks a ton! It works!
